### ISSUE TITLE
Pipes: | appends a newline to the selection, if necessary, while <a-|> does not.

### ISSUE THREAD

**Issue by bobi32** [mawww/kakoune#1216]
Yesterday I encountered some strange behavior while piping a selection into my clipboard with both these commands (long story short: `<a-|>` systematically lost the last character of the selection while `|` did not).
Then I led further tests and it turns out that `|` does append a newline to the selection, while `<a-|>` does not.

After a quick search on the repository, it seems that the issue comes from **src/normal.cc**. The git diff below implements a quick working patch:
```
diff --git a/src/normal.cc b/src/normal.cc
index 6a34f6a..db346cb 100644
--- a/src/normal.cc
+++ b/src/normal.cc
@@ -454,8 +454,11 @@ void pipe(Context& context, NormalParams)
                 for (int i = 0; i < selections.size(); ++i)
                 {
                     selections.set_main_index(i);
+                    String in = content(buffer, selections.main());
+                    if (in.back()!='\n')
+                        in += '\n';
                     ShellManager::instance().eval(real_cmd, context,
-                                                  content(buffer, selections.main()),
+                                                  in,
                                                   ShellManager::Flags::None);
                 }
             }
```

**Comment by mawww** [comment]
Hello,

I am not sure about this, in `|`, we assume the command is a filter, which usually wants full lines in the unix world. In the case we do add a trailing `\n`, we then remove it from the filter output.

For `<a-|>`, we dont know what the command is doing, and adding an eol is often a bad choice. The common use case for `<a-|>` is to pipe to `xsel` or `xclip` to copy the selection to X11 clipboard. Adding an eol there is clearly not the desired behaviour.

So, I dont think we can improve consistency between `|` and `<a-|>` by adding an eol in `<a-|>`, because `|` compensate that eol afterwards, `<a-|>` has no way to do that, and the eol is going to break the most common use case for that feature.

Out of curiosity, what was the use case that lead you to want this change ?

**Comment by bobi32** [comment]
Well, I started using `|` and `<a-|>` with X clipboard, but rather than using xsel or xclip, I used my clipboard manager (fyi, GPaste CLI tool `gpaste-client`: https://github.com/Keruspe/GPaste).

It is when I started to cut (with `|`, taking advantage that `gpaste-client`, with no further options, outputs nothing when reading from input pipe — just like `xsel -i` and `xclip -i` afaik) and copy (with `<a-|>`) selections that I noticed the different behaviors aforementioned.

When I saw the newline added in one case and not the other in **normal.cc**, I assumed that this was an issue for all CLI tools (xsel and xclip included), not only GPaste, my bad.

I may have a look at xsel or xclip for my bindings (replacing `|gpaste-client<ret>` with something like `<a-|>xsel -bi<ret>d`, in order to avoid some kind of reverse problem obviously). Thanks for the clarification about `|` and `<a-|>` semantics anyway!
